src/agent/src/image.rs

```diff
@@ -0,0 +1,323 @@
+const KATA_IMAGE_WORK_DIR: &str = "/run/image/";
+const CONFIG_JSON: &str = "config.json";
+const KATA_PAUSE_BUNDLE: &str = "/pause_bundle";
+
+const K8S_CONTAINER_TYPE_KEYS: [&str; 2] = [
+    "io.kubernetes.cri.container-type",
+    "io.kubernetes.cri-o.ContainerType",
+];
+
+#[rustfmt::skip]
+lazy_static! {
+    pub static ref IMAGE_SERVICE: Mutex<Option<ImageService>> = Mutex::new(None);
+}
+
+// Convenience function to obtain the scope logger.
+fn sl() -> slog::Logger {
+    slog_scope::logger().new(o!("subsystem" => "image"))
+}
+
+#[derive(Clone)]
+pub struct ImageService {
+    image_client: Arc<Mutex<ImageClient>>,
+    images: Arc<Mutex<HashMap<String, String>>>,
+}
+
+impl ImageService {
+    pub fn new() -> Self {
+        Self {
+            image_client: Arc::new(Mutex::new(ImageClient::new(PathBuf::from(
+                KATA_IMAGE_WORK_DIR,
+            )))),
+            images: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    /// Get the singleton instance of image service.
+    pub async fn singleton() -> Result<ImageService> {
+        IMAGE_SERVICE
+            .lock()
+            .await
+            .clone()
+            .ok_or_else(|| anyhow!("image service is uninitialized"))
+    }
+
+    async fn add_image(&self, image: String, cid: String) {
+        self.images.lock().await.insert(image, cid);
+    }
+
+    /// Set proxy environment from AGENT_CONFIG
+    fn set_proxy_env_vars() {
+        if env::var("HTTPS_PROXY").is_err() {
+            let https_proxy = &AGENT_CONFIG.https_proxy;
+            if !https_proxy.is_empty() {
+                env::set_var("HTTPS_PROXY", https_proxy);
+            }
+        }
+        if env::var("NO_PROXY").is_err() {
+            let no_proxy = &AGENT_CONFIG.no_proxy;
+            if !no_proxy.is_empty() {
+                env::set_var("NO_PROXY", no_proxy);
+            }
+        }
+    }
+
+    /// pause image is packaged in rootfs
+    fn unpack_pause_image(cid: &str, target_subpath: &str) -> Result<String> {
+        let cc_pause_bundle = Path::new(KATA_PAUSE_BUNDLE);
+        if !cc_pause_bundle.exists() {
+            bail!("Pause image not present in rootfs");
+        }
+
+        info!(sl(), "use guest pause image cid {:?}", cid);
+        let pause_bundle = Path::new(CONTAINER_BASE).join(cid).join(target_subpath);
+        let pause_rootfs = pause_bundle.join("rootfs");
+        fs::create_dir_all(&pause_rootfs)?;
+
+        let copy_if_not_exists = |src: &Path, dst: &Path| -> Result<()> {
+            if !dst.exists() {
+                info!(sl(), "copying file {src:?} to {dst:?}");
+                fs::copy(src, dst)?;
+            }
+            Ok(())
+        };
+        copy_if_not_exists(
+            &cc_pause_bundle.join(CONFIG_JSON),
+            &pause_bundle.join(CONFIG_JSON),
+        )?;
+        copy_if_not_exists(
+            &cc_pause_bundle.join("rootfs/pause"),
+            &pause_rootfs.join("pause"),
```

Comment:
We've just found an issue with this code: it makes the assumption that the pause image binary will be "pause", and located at the root of the container image. But we can have different images with different binaries.
For instance, the OCP pause image uses the binary "/usr/bin/pod", and not "pause".

If we embed a different pause image, the code need to be compatible. Maybe use the config.json file to find what command is to be used for that image, and copy the relevant binary instead of hardcoding it?

Reply:
Wow thanks for your suggestion @littlejawa. I overlook this situation. Let me see how to fix it.

Reply:
I dont know if this has been fixed with @ChengyuZhu6 latest changes, but I think we should not block this PR because of this perfectly valid report from @littlejawa .

I think we should go ahead and merge it and then open a new issue (potentially assigned to @ChengyuZhu6 ) to fix that problem. Would you be ok with that approach @littlejawa ?

Reply:
> I dont know if this has been fixed with @ChengyuZhu6 latest changes, but I think we should not block this PR because of this perfectly valid report from @littlejawa .
> 
> I think we should go ahead and merge it and then open a new issue (potentially assigned to @ChengyuZhu6 ) to fix that problem. Would you be ok with that approach @littlejawa ?

@sameo I fully agree with you,  that’s a good idea. WDYT, @littlejawa ?

Reply:
Sure, I don't want to block it. Just raised awareness to make sure we fix it.
We can go ahead for this PR.